File: C/tree/avltree/avltree.c
/**
 * AVL树 C语言版本
 */
#include <stdio.h>
#include <stdlib.h>
#include "avltree.h"

#define HEIGHT(p) ( (p == NULL) ? -1 : ((Node *)(p))->height )
#define MAX(a, b) ( (a) > (b) ? (a) : (b))

/**
 * 获取AVL树高度
 */
int avltree_height(AVLTree tree) {
    return HEIGHT(tree);
}

/**
 * 前序遍历AVL
 */
void preorder_avltree(AVLTree tree) {
    if(tree != NULL) {
        printf("%d ", tree->key);
        preorder_avltree(tree->left);
        preorder_avltree(tree->right);
    }
}

/**
 * 中序遍历AVL
 */
void inorder_avltree(AVLTree tree) {
    if(tree != NULL) {
        inorder_avltree(tree->left);
        printf("%d ", tree->key);
        inorder_avltree(tree->right);
    }
}

/**
 * 后序遍历AVL
 */
void postorder_avltree(AVLTree tree) {
    if(tree != NULL) {
        postorder_avltree(tree->left);
        postorder_avltree(tree->right);
        printf("%d ", tree->key);
    }
}

/**
 * (递归实现)查找“AVL树x"中键值为key的结点
 */ 
Node* avltree_search(AVLTree tree, TYPE key) {
    if(tree == NULL || tree->key == key) return tree;

    if(key < tree->key) {
        return avltree_search(tree->left, key);
    } else {
        return avltree_search(tree->right, key);
    }
}

/**
 * (非递归实现)查找”AVL树x“中键值为key的结点
 */
Node* iterative_avltree_search(AVLTree tree, TYPE key) {
    while(tree != NULL && tree->key != key) {
        if(key < tree->key) {
            tree = tree->left;
        } else {
            tree = tree->right;
        }
    }

    return tree;
}

/**
 * 查找最小结点： 返回tree为根结点的AVL树的最小结点
 */
Node* avltree_minimum(AVLTree tree) {
    if(tree == NULL) return NULL;

    while(tree->left != NULL) {
        tree = tree->left;
    }

    return tree;
}

/**
 * 查找最大结点：返回tree为根结点的AVL树的最大结点
 */
Node* avltree_maximum(AVLTree tree) {
    if(tree == NULL) return NULL;

    while(tree->right != NULL) {
        tree = tree->right;
    }

    return tree;
}

/**
 * LL : 左左对应的情况(左单旋转)
 * 返回值：旋转后的根结点
 */
static Node* left_left_rotation(AVLTree tree) {
    AVLTree retTree;

    retTree        = tree->left;
    tree->left     = retTree->right;
    retTree->right = tree;

    retTree->height = MAX(HEIGHT(retTree->left), HEIGHT(retTree->right)) + 1;
    tree->height    = MAX(HEIGHT(tree->left), HEIGHT(tree->right)) + 1;

    return retTree;
}

/**
 * RR : 右右对应的情况(右单旋转)
 * 返回值：旋转后的根结点
 */
static Node* right_right_rotation(AVLTree tree) {
    AVLTree retTree;

    retTree       = tree->right;
    tree->right   = retTree->left;
    retTree->left = tree;

    retTree->height = MAX(HEIGHT(retTree->left), HEIGHT(retTree->right)) + 1;
    tree->height    = MAX(HEIGHT(tree->left), HEIGHT(tree->right)) + 1;

    return retTree;
}

/**
 * LR ： 左右对应的情况(左双旋)
 * 返回值：旋转后的根结点
 */
static Node* left_right_rotation(AVLTree tree) {
    tree->left = right_right_rotation(tree->left);

    return left_left_rotation(tree);
}

/**
 * RL : 右左对应的情况(右双旋)
 * 返回值：旋转后的根结点
 */
static Node* right_left_rotation(AVLTree tree) {
    tree->right = left_left_rotation(tree->right);

    return right_right_rotation(tree);
}


/**
 * 创建AVL结点
 * 
 * --key   键值
 * --left  左孩子
 * --right 友孩子
 * 
 * 返回创建的结点
 */
static Node* avltree_create_node(TYPE key, Node* left, Node* right) {
    Node* p;

    if((p = (Node*)malloc(sizeof(Node))) == NULL) return NULL;

    p->key    = key;
    p->left   = left;
    p->right  = right;
    p->height = 0;

    return p;
};

/**
 * 将结点插入到树中，并返回根结点
 * 
 * -- tree AVL树根结点
 * -- key  插入结点的值
 * 
 * 返回插入后的树的根结点
 */
Node* avltree_insert(AVLTree tree, TYPE key) {
    if(tree == NULL) {
        // 创建新结点
        tree = avltree_create_node(key, NULL, NULL);
        if(tree == NULL) {
            printf("ERROR: create avltree node failed!\n");

            return NULL;
        }
    } else if(key < tree->key){
        // 将key插入到tree的左子树
        tree->left = avltree_insert(tree->left, key);
        // 插入接点后，若AVL树失去平衡,则进行相应的的调节
        if(HEIGHT(tree->left) - HEIGHT(tree->right) == 2) {
            if(key < tree->left->key) {
                tree = left_left_rotation(tree);
            } else {
                tree = left_right_rotation(tree);
            }
        }
    } else if(key > tree->key) {
        // 将key插入到tree的右子树
        tree->right = avltree_insert(tree->right, key);
        // 插入接点后，若AVL树失去平衡,则进行相应的的调节
        if(HEIGHT(tree->right) - HEIGHT(tree->left) == 2) {
            if(key < tree->right->key) {
                tree = right_left_rotation(tree);
            } else {
                tree = right_right_rotation(tree);
            }
        }
    } else {
        printf("添加失败: 不允许添加相同的结点!\n");
    }

    tree->height = MAX(HEIGHT(tree->left), HEIGHT(tree->right)) + 1;

    return tree;
}

/**
 * 删除结点node, 返回根结点
 * 
 * -- tree AVL树根结点
 * -- node 待删除的结点
 * 
 * 返回删除后的根结点
 */
static Node* delete_node(AVLTree tree, Node* node){
    // 如果根为或没有待删除的结点，直接返回
    if (tree == NULL || node == NULL) return NULL;

    if(node->key < tree->key) {
        // 待删除的结点在左子树中
        tree->left = delete_node(tree->left, node);
        // 删除结点后，AVL树失去平衡，则进相应的调节
        if(HEIGHT(tree->right) - HEIGHT(tree->left) == 2) {
            Node* right = tree->right;
            if(HEIGHT(right->left) > HEIGHT(right->right)) {
                tree = right_left_rotation(tree);
            } else {
                tree = right_right_rotation(tree);
        }
    } else if(node->key > tree->key) {
        // 待删除的结点在右子树中
        tree->right = delete_node(tree->right, node);
        // 删除结点后，AVL树失去平衡，则进相应的调节
        if(HEIGHT(tree->left) - HEIGHT(tree->right) == 2) {
            Node* left = tree->left;
            if(HEIGHT(left->left) > HEIGHT(left->right)) {
                tree = left_left_rotation(tree);
            } else {
                tree = left_right_rotation(tree);
            }
        }
    } else {
        // tree是对应要删除的结点
        if((tree->left) && (tree->right)) {
            if(HEIGHT(tree->left) > HEIGHT(tree->right)) {
                // 如果tree的左子树比右子树高；
                // 则(01)找出tree的左子树中的最大节点
                //   (02)将该最大节点的值赋值给tree。
                //   (03)删除该最大节点。
                // 这类似于用"tree的左子树中最大节点"做"tree"的替身；
                // 采用这种方式的好处是：删除"tree的左子树中最大节点"之后，AVL树仍然是平衡的
                Node* max  = avltree_maximum(tree->left);
                tree->key  = max->key;
                tree->left = delete_node(tree->left, max);
            } else {
                // 如果tree的左子树不比右子树高(即它们相等，或右子树比左子树高1)
                // 则(01)找出tree的右子树中的最小节点
                //   (02)将该最小节点的值赋值给tree。
                //   (03)删除该最小节点。
                // 这类似于用"tree的右子树中最小节点"做"tree"的替身；
                // 采用这种方式的好处是：删除"tree的右子树中最小节点"之后，AVL树仍然是平衡的。
                Node* min   = avltree_minimum(tree->right);
                tree->key   = min->key;
                tree->right = delete_node(tree->right, min);
            }
        } else {
            Node* temp = tree;
            tree = tree->left ? tree->left : tree->right;
            free(temp);
            }
        }
    }

    if(tree != NULL) {
        tree->height = MAX(HEIGHT(tree->left), HEIGHT(tree->right)) + 1;
    }

    return tree;
}

/**
 * 删除值为key的结点，返回根结点
 * 
 * -- tree AVL树的根结点
 * -- key  删除结点的值
 * 
 * 返回删除后的根结点
 */
Node* avltree_delete(AVLTree tree, TYPE key) {
    Node* node;

    if((node = avltree_search(tree, key)) != NULL) tree = delete_node(tree, node);

    return tree;
}

/**
 * 摧毁AVL树
 */
void destroy_avltree(AVLTree tree) {
    if(tree == NULL) return;

    if(tree->left != NULL) {
        destroy_avltree(tree->left);
    }

    if(tree->right != NULL) {
        destroy_avltree(tree->right);
    }

    free(tree);
}

/**
 * 打印AVL树
 * 
 * -- tree       树的根结点
 * -- key        结点值
 * -- direction  0: 表示该结点是根
 *              -1：表示该结点是它父结点的左孩子
 *               1：表示该结点是它父结点的右孩子
 */
void print_avltree(AVLTree tree, TYPE key, int direction) {
    if(tree != NULL) {
        if(direction == 0) {
            printf("%2d is root\n", key);
        } else {
            printf("%2d is %2d's %6s child\n", tree->key, key, direction == 1 ? "right" : "left");
        }

        print_avltree(tree->left, tree->key, -1);
        print_avltree(tree->right, tree->key, 1);
    }
}

File: JAVA/tree/avltree/AvlTree.java
package tree.avltree;

/**
 * AVL树Java实现
 */

public class AvlTree<T extends Comparable<T>> {
    AvlTreeNode<T> root;

    class AvlTreeNode<T extends Comparable<T>> {
        T   key;    // 结点值
        int height; // 高度
        
        AvlTreeNode<T> left; // 左孩子
        AvlTreeNode<T> right; // 右孩子

        public AvlTreeNode(final T key, final AvlTreeNode<T> left, final AvlTreeNode<T> right) {
            this.key    = key;
            this.left   = left;
            this.right  = right;
            this.height = 0;
        }
    }

    public AvlTree() {
        this.root = null;
    }

    /**
     * 获取树的高度
     */
    private int height(final AvlTreeNode<T> node) {
        return node != null ? node.height : 0;
    }

    /**
     * 树的高度
     */
    public int height() {
        return height(root);
    }

    /**
     * 比较两个值的大小
     */
    private int max(final int a, final int b) {
        return a > b ? a : b;
    }

    /**
     * 前序遍历
     */
    private void preOrder(final AvlTreeNode<T> tree) {
        if(tree != null) {
            System.out.print(tree.key + " ");
            preOrder(tree.left);
            preOrder(tree.right);
        }
    }

    public void preOrder() {
        preOrder(root);
    }

    /**
     * 中序遍历
     */
    private void inOrder(final AvlTreeNode<T> tree) {
        if(tree != null) {
            inOrder(tree.left);
            System.out.print(tree.key + " ");
            inOrder(tree.right);
        }
    }

    public void inOrder() {
        inOrder(root);
    }

    /**
     * 后序遍历
     */
    private void postOrder(final AvlTreeNode<T> tree) {
        if(tree != null) {
            postOrder(tree.left);
            postOrder(tree.right);
            System.out.print(tree.key + " ");
        }
    }

    public void postOrder() {
        postOrder(root);
    }

    /**
     * 查找tree 树中值为key的结点(递归实现)
     * 
     * @param  tree avl树根结点
     * @param  key  查找的值
     * @return
     */
    private AvlTreeNode<T> search(final AvlTreeNode<T> tree, final T key) {
        if(tree == null) return null;

        if(key.compareTo(tree.key) < 0) {
            return search(tree.left, key);
        } else if(key.compareTo(tree.key) > 0) {
            return search(tree.right, key);
        } else {
            return tree;
        }
    }

    public AvlTreeNode<T> search(final T key) {
        return search(root, key);
    }

    /**
     * 查找tree 树中值为key的结点(递归实现
     * 
     * @param  tree avl树根结点
     * @param  key  查找的值
     * @return
     */
    private AvlTreeNode<T> iteratorSearch(AvlTreeNode<T> tree, final T key) {
        while(tree != null) {
            if(key.compareTo(tree.key) > 0) {
                tree = tree.right;
            } else if(key.compareTo(tree.key) < 0) {
                tree =tree.left;
            } else {
                return tree;
            }
        }
        
        return tree;
    }

    public AvlTreeNode<T> iteratorSearch(final T key) {
        return iteratorSearch(root, key);
    }

    /**
     * 查找tree树中的最小值
     * 
     * @param  tree 查找最小树的值
     * @return
     */
    private AvlTreeNode<T> minimum(AvlTreeNode<T> tree) {
        if(tree == null) return null;
        
        while(tree.left != null) {
            tree = tree.left;
        }

        return tree;
    }

    public T minimum() {
        AvlTreeNode<T> min = minimum(root);
        return min != null ? min.key : null;
    }

    /**
     * 查找tree树中的最大值
     * 
     * @param   tree 查找最大树的值
     * @return
     */
    private AvlTreeNode<T> maximum(AvlTreeNode<T> tree) {
        if(tree == null) return null;

        while(tree.right != null) {
            tree = tree.right;
        }

        return tree;
    }

    public T maximum() {
        AvlTreeNode<T> max = maximum(root);
        return max != null ? max.key : null;
    }

    /**
     * LL: 左左对应的情况(左单旋转)
     * 
     * @param  tree 进行旋转的tree
     * @return
     */
    private AvlTreeNode<T> leftLeftRotation(final AvlTreeNode<T> tree) {
        AvlTreeNode<T> tmp = tree.left;

        tree.left = tmp.right;
        tmp.right = tree;

        tree.height = max(height(tree.left), height(tree.right)) + 1;
        tmp.height  = max(height(tmp.left), height(tmp.right)) + 1;

        return tmp;
    }

    /**
     * RR: 右右对应的情况(右单旋转)
     * 
     * @param  tree 进行旋转的tree
     * @return
     */
    private AvlTreeNode<T> rightRightRotation(final AvlTreeNode<T> tree) {
        AvlTreeNode<T> tmp = tree.right;

        tree.right = tmp.left;
        tmp.left   = tree;

        tree.height = max(height(tree.left), height(tree.right)) + 1;
        tmp.height  = max(height(tmp.left), height(tree.right)) + 1;

        return tmp;
    }

    /**
     * LR: 左右对应的情况(左双旋转)
     * 
     * @param  tree 进行旋转的tree
     * @return
     */
    private AvlTreeNode<T> leftRightRotation(final AvlTreeNode<T> tree) {
        tree.left = rightRightRotation(tree.left);

        return leftLeftRotation(tree);
    }

    /**
     * RL: 右左对应的情况(右双旋转)
     * 
     * @param  tree 进行旋转的tree
     * @return
     */
    private AvlTreeNode<T> rightLeftRotation(final AvlTreeNode<T> tree) {
        tree.right = leftLeftRotation(tree.right);

        return rightRightRotation(tree);
    }    

    /**
     * 向AVL树中插入一个新值Key
     * 
     * @param  tree 插入的值的树
     * @param  key  插入的值
     * @return
     */
    private AvlTreeNode<T> insert(AvlTreeNode<T> tree, T key) {
        if(tree == null) {
            tree = new AvlTreeNode<T>(key, null, null);
        } else if(key.compareTo(tree.key) < 0) {
            // 插入值应该在tree的左孩子中
            tree.left = insert(tree.left, key);
            // 插入值如果影响tree的平衡就需要进行恢复平衡操作
            if(height(tree.left) - height(tree.right) == 2) {
                if(key.compareTo(tree.left.key) > 0) {
                    tree = leftRightRotation(tree);
                } else {
                    tree = leftLeftRotation(tree);
                }
            }
        } else if(key.compareTo(tree.key) > 0) {
            // 插入值应该在tree的右孩子中
            tree.right = insert(tree.right, key);
            // 插入值如果影响tree的平衡就需要进行恢复平衡操作
            if(height(tree.right) - height(tree.left) == 2) {
                if(key.compareTo(tree.right.key) > 0) {
                    tree = rightRightRotation(tree);
                } else {
                    tree = rightLeftRotation(tree);
                }
            }
        } else {
            System.out.println("插入新元素错误，不能存在相同的数");
        }

        tree.height = max(height(tree.left), height(tree.right)) + 1;

        return tree;
    }

    public void insert(T key) {
        root = insert(root, key);
    }

    /**
     * 从平衡树中删除结点key
     * 
     * @param  tree 删除的平衡树
     * @param  key  删除的结点
     * @return
     */
    private AvlTreeNode<T> remove(AvlTreeNode<T> tree, final AvlTreeNode<T> key) {
        if(tree == null || key == null) return null;

        if(key.key.compareTo(tree.key) < 0) {
            // 删除值应该发生在左孩子中
            tree.left = remove(tree.left, key);
            // 删除值如果影响tree的平衡就需要进行恢复平衡操作
            if(height(tree.right) - height(tree.left) == 2) {
                if(key.key.compareTo(tree.left.key) < 0) {
                    final AvlTreeNode<T> right = tree.right;
                    if(height(right.left) > height(right.right)) {
                        tree = rightLeftRotation(tree);
                    } else {
                        tree = rightRightRotation(tree);
                    }
                }
            }
        } else if(key.key.compareTo(tree.key) > 0){
            // 删除的值应该发生在右孩子的中
            tree.right = remove(tree.right, key);
            // 删除值如果影响tree的平衡就需要进项恢复平衡操作
            if(height(tree.left) - height(tree.right) == 2) {
                final AvlTreeNode<T> left = tree.left;
                if(height(left.left) > height(left.right)) {
                    tree = leftLeftRotation(tree);
                } else {
                    tree = leftRightRotation(tree);
                }
            }
        } else {
            // 要删除的key结点就是tree
            if(tree.left != null && tree.right != null) {
                if(height(tree.left) > height(tree.right)) {
                    // 如果tree的左子树比右子树高；
                    // 则(01)找出tree的左子树中的最大节点
                    //   (02)将该最大节点的值赋值给tree。
                    //   (03)删除该最大节点。
                    // 这类似于用"tree的左子树中最大节点"做"tree"的替身；
                    // 采用这种方式的好处是：删除"tree的左子树中最大节点"之后，AVL树仍然是平衡的
                    final AvlTreeNode<T> maximum = maximum(tree.left);
                    tree.key = maximum.key;
                    tree.left = remove(tree.left, maximum);
                } else {
                    // 如果tree的左子树不比右子树高(即它们相等，或右子树比左子树高1)
                    // 则(01)找出tree的右子树中的最小节点
                    //   (02)将该最小节点的值赋值给tree。
                    //   (03)删除该最小节点。
                    // 这类似于用"tree的右子树中最小节点"做"tree"的替身；
                    // 采用这种方式的好处是：删除"tree的右子树中最小节点"之后，AVL树仍然是平衡的
                    final AvlTreeNode<T> minimum = minimum(tree.right);
                    tree.key = minimum.key;
                    tree.right = remove(tree.right, minimum);
                }
            } else {
                AvlTreeNode<T> tmp = tree;
                tree = tree.left != null ? tree.left : tree.right;
                tmp  = null;
            }
        }

        if(tree != null) {
            tree.height = max(height(tree.left), height(tree.right)) + 1;
        }

        return tree;
    }

    public void remove(final T key) {
        final AvlTreeNode<T> node = search(key);
        if(node != null){
            remove(root, node);
        }
    }

    /**
     * 摧毁avl平衡树
     * 
     * @param tree
     */
    private void destory(AvlTreeNode<T> tree) {
        if(tree == null) return;

        if(tree.left != null) {
            destory(tree.left);
        }
        if(tree.right != null) {
            destory(tree.right);
        }

        tree = null;
    }

    public void destory() {
        destory(root);
    }

    /**
     * 打印 AVL树
     *  
     * @param tree        树
     * @param key         父结点值
     * @param direction   0表示根/-1表示左孩子/1表示右孩子
     */
    private void print(final AvlTreeNode<T> tree, final T key, final int direction) {
        if(tree != null) {
            if(direction == 0) {
                System.out.printf("%2d is root\n", tree.key);
            } else {
                System.out.printf("%2d is %2d's %6s child\n", tree.key, key, direction == 1 ? "left" : "right");
            }

            print(tree.left, tree.key, -1);
            print(tree.right, tree.key, 1);
        }
    }

    public void print() {
        if(root != null) {
            print(root, root.key, 0);
        }
    }
}